src/main/java/com/lightbend/akka/jpa/sample/UserRepository.java

```diff
@@ -0,0 +1,140 @@
+
+    } catch (NoResultException e) {
+      return Optional.empty();
+    }
+  }
+
+  public CompletionStage<List<User>> findAll() {
+    return withEntityManager(entityManager -> {
+      CriteriaBuilder builder = entityManager.getCriteriaBuilder();
+      CriteriaQuery<User> criteria = builder.createQuery(User.class);
+      Root<User> root = criteria.from(User.class);
+      criteria.select(root);
+
+      List<User> result = entityManager.createQuery(criteria).getResultList();
+      return result;
+    });
+  }
+
+  public CompletionStage<List<User>> findByName(String name) {
+    return withEntityManager(entityManager -> {
+      CriteriaBuilder builder = entityManager.getCriteriaBuilder();
+      CriteriaQuery<User> criteria = builder.createQuery(User.class);
+      Root<User> root = criteria.from(User.class);
+      criteria.select(root).where(builder.equal(root.get("name"), name));
+
+      List<User> result = entityManager.createQuery(criteria).getResultList();
+      return result;
+    });
+  }
+
+  public CompletionStage<Done> save(User user) {
+    return withEntityManager(entityManager -> {
+      if (user.isNew()) {
+        // new user
+        entityManager.persist(user);
+      } else {
+        // update
+        entityManager.merge(user);
+      }
+      return Done.getInstance();
+    });
+  }
+
+  public CompletionStage<Done> delete(long id) {
+    return withEntityManager(entityManager -> {
+      Optional<User> maybeUser = blockingFindById(entityManager, id);
+      if (maybeUser.isPresent())
+        entityManager.remove(maybeUser.get());
+      return Done.getInstance();
+    });
+  }
+
+  /**
+   * All data access should be performed via this method.
+   *
+   * It handles the lifecycle of the EntityManager, i.e. closing it after
+   * the data access code returns or throws.
+   *
+   * A transaction is started before the data access code is executed and
+   * it is committed afterwards  or rolled back if exception was thrown.
+   *
+   * The blocking data access code is executed in a dedicated Akka dispatcher
+   * (thread pool) to avoid starvation of other non-blocking parts of the system.
+   *
+   * The `CompletionStage` is returned immediately and completed later when with the
+   * value that the data access code returns. If exception is thrown by the entity manager,
+   * the data access code, or transaction the `CompletionStage` is completed exceptionally
+   * with that exception.
+   */
+  private <T> CompletionStage<T> withEntityManager(Function<EntityManager, T> dataAccess) {
+    CompletableFuture<T> result = new CompletableFuture<>();
+    blockingDispatcher.execute(() -> {
+      try {
+        final EntityManager entityManager = entityManagerFactory.createEntityManager();
+        try {
+          entityManager.getTransaction().begin();
+          T r = dataAccess.apply(entityManager);
+          entityManager.getTransaction().commit();
+          entityManager.close();
```

Comment:
Shouldn't this be in a `finally` block?

Reply:
The reason I placed it here and inside the catch, is that in the failure case I also try to close it but if the close but then I'm interested of the original exception and not of the one that could potentially be thrown by `close`.